modules/seqexec/engine/src/main/scala/seqexec/engine/Engine.scala

```diff
@@ -278,36 +204,36 @@ class Engine[D, U](stateL: Lens[D, Engine.State]) {
     /**
       * Log info lifted into Handle.
       */
-    def info(msg: => String): HandleP[Unit] = pure((logger.info(msg), None)).void
+    def info(msg: => String): HandleType[Unit] = pure((logger.info(msg), None)).void
 
     /**
       * Log warning lifted into Handle.
       */
-    def warning(msg: => String): HandleP[Unit] = pure((logger.warn(msg), None)).void
+    def warning(msg: => String): HandleType[Unit] = pure((logger.warn(msg), None)).void
 
     /**
       * Log debug lifted into Handle.
       */
-    def debug(msg: => String): HandleP[Unit] = pure((logger.debug(msg), None)).void
+    def debug(msg: => String): HandleType[Unit] = pure((logger.debug(msg), None)).void
 
     /**
       * Log error lifted into Handle
       */
-    def error(msg: => String): HandleP[Unit] = pure((logger.error(msg), None)).void
+    def error(msg: => String): HandleType[Unit] = pure((logger.error(msg), None)).void
 
   }
 
   /**
     * Enqueue `Event` in the Handle.
     */
-  private def send(ev: EventType): HandleP[Unit] = HandleP.fromStream(Stream(ev))
+  private def send(ev: EventType): HandleType[Unit] = HandleP.fromStream(Stream(ev))
 
   /**
     * Main logical thread to handle events and produce output.
     */
   @SuppressWarnings(Array("org.wartremover.warts.ImplicitParameter"))
-  private def run(ev: EventType)(implicit ec: ExecutionContext): HandleP[ResultType] = {
-    def handleUserEvent(ue: UserEventType): HandleP[ResultType] = ue match {
+  private def run(userReact: PartialFunction[SystemEvent, HandleType[Unit]])(ev: EventType)(implicit ec: ExecutionContext): HandleType[ResultType] = {
```

Comment:
Now the user of `Engine` can pass a function to react to system events inside the engine loop.